hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/qjournal/TestMiniJournalCluster.java

```diff
@@ -52,4 +53,70 @@ public void testStartStop() throws IOException {
       c.shutdown();
     }
   }
+
+  @Test
+  public void testStartStopWithPorts() throws IOException {
+    Configuration conf = new Configuration();
+
+    try {
+      new MiniJournalCluster.Builder(conf).setHttpPorts(8481).build();
+      fail("Should not reach here");
+    } catch (IllegalArgumentException e) {
+      assertEquals("Num of http ports (1) should match num of JournalNodes (3)", e.getMessage());
+    }
+
+    try {
+      new MiniJournalCluster.Builder(conf).setRpcPorts(8481, 8482)
+          .build();
+      fail("Should not reach here");
+    } catch (IllegalArgumentException e) {
+      assertEquals("Num of rpc ports (2) should match num of JournalNodes (3)", e.getMessage());
+    }
+
+    try {
+      new MiniJournalCluster.Builder(conf).setHttpPorts(800, 9000, 10000).setRpcPorts(8481)
+          .build();
+      fail("Should not reach here");
+    } catch (IllegalArgumentException e) {
+      assertEquals("Num of rpc ports (1) should match num of JournalNodes (3)", e.getMessage());
+    }
+
+    try {
+      new MiniJournalCluster.Builder(conf).setHttpPorts(800, 9000, 1000, 2000)
+          .setRpcPorts(8481, 8482, 8483)
+          .build();
+      fail("Should not reach here");
+    } catch (IllegalArgumentException e) {
+      assertEquals("Num of http ports (4) should match num of JournalNodes (3)", e.getMessage());
+    }
+
+    MiniJournalCluster miniJournalCluster =
+        new MiniJournalCluster.Builder(conf).setHttpPorts(8481, 8482, 8483)
+            .setRpcPorts(8491, 8492, 8493).build();
```

Comment:
@tomscut this is the only UT that will use `setHttpPorts` and `setRpcPorts` to set custom ports and then asserts that correct ports are used, hence this UT should not collide with any other UTs. Besides, we must have some UT to ensure both `setHttpPorts` and `setRpcPorts` are working as expected, correct?
Thanks

Reply:
> hence this UT should not collide with any other UTs.

I have similar concerns, We are hard coding the ports here, this test can fail in specific environments where these ports are already occupied. We have seen such cases in past..
--> Just had a cursory look

Reply:
Thanks. @ayushtkn I just wanted to see if this UT with hardcoded ports pass in Jenkins build and since it did, is it good enough for us to keep? If not, it's fine and I can try to randomize the ports generation (similar to default case) and assert that randomly chosen port is the one being used by MiniJournal cluster.

Reply:
We can not rely completely on Jenkins, I remember seeing an issue with a RBF test, where we had a port hard-coded, and for some organisation, there internal build was failing for that test, because they had something running on that port.

May be first finding some free ports, and then putting them into this conf should do. Or if that isn't possible, second option is try some randomisation, get a set of some random ports in a specified range, if they work great, if not loop back find another set and so on for some specified amount of iterations.

If above two doesn't work, we can try some skipping mechanism, like ports are occupied so skip the test or so, But this would be the last and the worst thing to do.

And in Jenkins the tests run in parallel, so results might change depending on what tests are running together & what ports they randomly choose. 

My general experience so far, such controversial tests don't fail in the actual PR, but in other folks PR. :-P 
Better we  play safe. :)

Reply:
> We can not rely completely on Jenkins, I remember seeing an issue with a RBF test, where we had a port hard-coded, and for some organisation, there internal build was failing for that test, because they had something running on that port.

Valid point!